ddl/cache.lua

```diff
@@ -0,0 +1,136 @@
+local fiber = require('fiber')
+
+local cache = nil
+
+local CACHE_LOCK_TIMEOUT  = 3
+local SPACE_NAME_IDX      = 1
+local SHARD_FUNC_NAME_IDX = 2
+local SHARD_FUNC_BODY_IDX = 3
+
+-- Function decorator that is used to prevent cache_build() from being
+-- called concurrently by different fibers.
+local function locked(f)
+    return function(...)
+        local ok = cache.lock:put(true, CACHE_LOCK_TIMEOUT)
+
+        if not ok then
+            error("cache lock timeout is exceeded")
+        end
+
+        local status, err = pcall(f, ...)
+        cache.lock:get()
+
+        if not status or err ~= nil then
+            return err
+        end
+    end
+end
+
+-- Build cache.
+--
+-- Cache structure format:
+-- cache = {
+--   spaces = {
+--     'space_name' = {
+--       raw = {},       -- raw sharding metadata, used for ddl.get()
+--       processed = {}, -- table with parsed dot notation (like {'foo', 'bar'})
+--                       -- or a function ready to call (or a string with an error)
+--     }
+--   },
+--   lock,               -- locking based on fiber.channel()
+--   schema_version,     -- current schema version
+-- }
+--
+-- function returns nothing
+local cache_build = locked(function()
+    -- clear cache
+    cache.spaces = {}
+
+    if box.space._ddl_sharding_func == nil then
+        return
+    end
+
+    for _, tuple in box.space._ddl_sharding_func:pairs() do
+        local space_name = tuple[SPACE_NAME_IDX]
+        local func_name = tuple[SHARD_FUNC_NAME_IDX]
+        local func_body = tuple[SHARD_FUNC_BODY_IDX]
+
+        cache.spaces[space_name] = {
+            raw = tuple
+        }
+
+        if func_body ~= nil then
+            local sharding_func, err = loadstring('return ' .. func_body)
+            if sharding_func == nil then
+                cache.spaces[space_name].processed =
+                    string.format("Body is incorrect in sharding_func for space (%s): %s",
+                    space_name, err)
+            else
+                cache.spaces[space_name].processed =
+                    sharding_func()
+            end
+        elseif func_name ~= nil then
+            local chunks = string.split(func_name, '.')
+            cache.spaces[space_name].processed = chunks
```

Comment:
Nit: As for me, it requires a clarification: why we don't save the function itself into the cache? I know, it is because the function can be changed in runtime and we have no way to catch this change, but anyway: another reader may be confused.

Reply:
comment added